src/db.c

```diff
@@ -4128,206 +4155,1718 @@ db_speaker_clear_all(void)
+db_queue_get_version()
 {
-#define Q_TMPL "INSERT INTO inotify (wd, cookie, path) VALUES (%d, 0, '%q');"
-  char *query;
+  char *version;
+  int32_t plversion;
+  int ret;
 
-  query = sqlite3_mprintf(Q_TMPL, wi->wd, wi->path);
+  plversion = 0;
+  version = db_admin_get("plversion");
+  if (version)
+    {
+      ret = safe_atoi32(version, &plversion);
+      free(version);
+      if (ret < 0)
+	{
+	  DPRINTF(E_LOG, L_DB, "Could not get playlist version\n");
+	  return -1;
+	}
+    }
 
-  return db_query_run(query, 1, 0);
-#undef Q_TMPL
+  return plversion;
 }
 
-int
-db_watch_delete_bywd(uint32_t wd)
+/*
+ * Increments the version of the queue in the admin table and notifies listener of LISTENER_PLAYLIST
+ * about the change.
+ *
+ * This function must be called after successfully modifying the queue table in order to send
+ * notification messages to the clients (e. g. dacp or mpd clients).
+ */
+static void
+queue_inc_version_and_notify()
 {
-#define Q_TMPL "DELETE FROM inotify WHERE wd = %d;"
-  char *query;
+  int plversion;
+  char version[10];
+  int ret;
 
-  query = sqlite3_mprintf(Q_TMPL, wd);
+  db_transaction_begin();
 
-  return db_query_run(query, 1, 0);
-#undef Q_TMPL
-}
+  plversion = db_queue_get_version();
+  if (plversion < 0)
+    plversion = 0;
 
-int
-db_watch_delete_bypath(char *path)
-{
-#define Q_TMPL "DELETE FROM inotify WHERE path = '%q';"
-  char *query;
+  plversion++;
+  ret = snprintf(version, sizeof(version), "%d", plversion);
+  if (ret >= sizeof(version))
+    {
+      DPRINTF(E_LOG, L_DB, "Error incrementing queue version. Could not convert version to string: %d\n", plversion);
+      db_transaction_rollback();
+      return;
+    }
 
-  query = sqlite3_mprintf(Q_TMPL, path);
+  ret = db_admin_update("plversion", version);
+  if (ret < 0)
+    {
+      DPRINTF(E_LOG, L_DB, "Error incrementing queue version. Could not update version in admin table: %d\n", plversion);
+      db_transaction_rollback();
+      return;
+    }
 
-  return db_query_run(query, 1, 0);
-#undef Q_TMPL
+  db_transaction_end();
+
+  listener_notify(LISTENER_PLAYLIST);
 }
 
-int
-db_watch_delete_bymatch(char *path)
+void
+db_queue_update_icymetadata(int id, char *artist, char *album)
 {
-#define Q_TMPL "DELETE FROM inotify WHERE path LIKE '%q/%%';"
+#define Q_TMPL "UPDATE queue SET artist = TRIM(%Q), album = TRIM(%Q) WHERE id = %d;"
   char *query;
 
-  query = sqlite3_mprintf(Q_TMPL, path);
+  if (id == 0)
+    return;
+
+  query = sqlite3_mprintf(Q_TMPL, artist, album, id);
+  if (!query)
+    {
+      DPRINTF(E_LOG, L_DB, "Out of memory for query string\n");
+
+      return;
+    }
+
+  db_query_run(query, 1, 0);
+  queue_inc_version_and_notify();
 
-  return db_query_run(query, 1, 0);
 #undef Q_TMPL
 }
 
-int
-db_watch_delete_bycookie(uint32_t cookie)
-{
-#define Q_TMPL "DELETE FROM inotify WHERE cookie = %" PRIi64 ";"
+static int
+queue_add_mediafileinfo(struct db_media_file_info *dbmfi, int pos, int shuffle_pos)
```

Comment:
Would renaming it to queue_add_file be ok? (i think the "db_" prefix is used to mark public functions)